src/storage/src/window_infer.rs

```diff
@@ -0,0 +1,328 @@
+use common_time::timestamp::TimeUnit;
+use common_time::timestamp_millis::BucketAligned;
+use itertools::Itertools;
+
+use crate::memtable::MemtableStats;
+use crate::sst::FileMeta;
+
+/// A set of predefined time windows.
+const TIME_WINDOW_SIZE: [i64; 9] = [
+    60,               // 1 minute
+    60 * 10,          // 10 minute
+    60 * 30,          // 30 minute
+    60 * 60,          // 1 hour
+    2 * 60 * 60,      // 2 hours
+    6 * 60 * 60,      // 6 hours
+    12 * 60 * 60,     // 12 hours
+    24 * 60 * 60,     // 1 day
+    7 * 24 * 60 * 60, // 1 week
+];
+
+/// [WindowInfer] infers the time windows that can be used to optimize table scans ordered by
+/// timestamp column or have explicit time windows. By splitting time spans of tables into
+/// time windows, we can scan entries window by window.
+pub(crate) trait WindowInfer {
+    /// Infers time windows according to the SST files and memtables.
+    ///
+    /// ### Note
+    /// The order of returned vector defines how records are yielded.
+    fn infer_window(&self, files: &[FileMeta], mem_tables: &[MemtableStats])
+        -> Vec<TimestampRange>;
+}
+
+/// [PlainWindowInference] simply finds the minimum time span within all SST files in level 0 and
+/// memtables, matches that time span into a set of predefined time windows.
+pub(crate) struct PlainWindowInference;
+
+impl WindowInfer for PlainWindowInference {
+    fn infer_window(
+        &self,
+        files: &[FileMeta],
+        mem_tables: &[MemtableStats],
+    ) -> Vec<TimestampRange> {
+        let mut min_duration_sec = i64::MAX;
+        let mut durations = Vec::with_capacity(files.len() + mem_tables.len());
+
+        for meta in files {
+            if let Some((start, end)) = &meta.time_range {
+                // unwrap safety: converting timestamps with any unit to seconds won't overflow.
+                let start_sec = start.convert_to(TimeUnit::Second).unwrap().value();
+                // file timestamp range end is inclusive
+                let end_sec = end.convert_to_ceil(TimeUnit::Second).unwrap().value();
+                debug_assert!(end_sec >= start_sec);
+                if meta.level == 0 {
+                    // only level 0 is involved when calculating time windows.
+                    min_duration_sec = min_duration_sec.min(end_sec - start_sec);
+                }
+                durations.push((start_sec, end_sec));
+            }
+        }
+
+        for stats in mem_tables {
+            // unwrap safety: converting timestamps with any unit to seconds won't overflow.
+            let start_sec = stats
+                .min_timestamp
+                .convert_to(TimeUnit::Second)
+                .unwrap()
+                .value();
+            let end_sec = stats
+                .max_timestamp
+                .convert_to_ceil(TimeUnit::Second)
+                .unwrap()
+                .value();
+            min_duration_sec = min_duration_sec.min(end_sec - start_sec);
```

Comment:
will this generate a duration with `0s`? looks like the following tests don't cover this case

Reply:
Not actually since min_timestamp is converted to `TimeUnit::Second` by `Timestamp::convert_to` while max_timestamp is converted by `Timestamp::convert_to_ceil`. 

Even if `min_duration_sec` is zero, it will also be aligned to 60 seconds.